[RPC profiling] Don't wrap toHere() calls with profiling

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44967 Fallback to CPU when remote end does not have CUDA for profiling
* #44923 [RPC profiling] Add tests to ensure RPC profiling works on single threaded
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.

Since `toHere()` does not execute operations (torch operators) over RPC and simply
transfers the value to the local node, we don't need to enable the profiler
remotely for this message. This causes unnecessary overhead and is not needed.

Since `toHere` is a blocking call, we already profile the call on the local node using `RECORD_USER_SCOPE`, so this does not change the expected profiler results (validated by ensuring all remote profiling tests pass).

Differential Revision: [D23641466](https://our.internmc.facebook.com/intern/diff/D23641466/)

rohan-varma committed:
[RPC profiling] Don't wrap toHere() calls with profiling

Since `toHere()` does not execute operations over RPC and simply
transfers the value to the local node, we don't need to enable the profiler
remotely for this message. This causes unnecessary overhead and is not needed.

Since `toHere` is a blocking call, we already profile the call on the local node using `RECORD_USER_SCOPE`, so this does not change the expected profiler results (validated by ensuring all remote profiling tests pass).

Differential Revision: [D23641466](https://our.internmc.facebook.com/intern/diff/D23641466/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/autograd/utils.cpp:
```diff
@@ -153,7 +154,7 @@ std::shared_ptr<FutureMessage> sendMessageWithAutograd(
   // If profiler is enabled, wrap this message with profiling metadata that will
   // tell the remote end to process this request with the profiler enabled.
-  if (torch::autograd::profiler::profilerEnabled()) {
+  if (!forceDisableProfiling && torch::autograd::profiler::profilerEnabled()) {
```
This looks OK to me. But I wonder if we can avoid this new arg by making `torch::autograd::profiler::profilerEnabled()` return false? E.g., is it possible to use a guard in `rref_impl.cpp` to toggle the status of `profilerEnabled`?

rohan-varma replied:
I guess one thing we could do is set a thread_local that allows this override, and have a guard that sets/restores it. If it's true, then we would override `profilerEnabled()` to always be false. Although, I don't think this should exist in `profiler::profilerEnabled()` itself since that may increase complexity in the profiler too much, maybe we can have a separate function for RPC profiling specifically that wraps around `profilerEnabled()` and checks this guard value.

mrshenli replied:
I see, if it's not that easy to toggle profiler on/off. The current version LGTM! Thanks!

rohan-varma committed:
Update on "[RPC profiling] Don't wrap toHere() calls with profiling"


Since `toHere()` does not execute operations (torch operators) over RPC and simply
transfers the value to the local node, we don't need to enable the profiler
remotely for this message. This causes unnecessary overhead and is not needed.

Since `toHere` is a blocking call, we already profile the call on the local node using `RECORD_USER_SCOPE`, so this does not change the expected profiler results (validated by ensuring all remote profiling tests pass).

Differential Revision: [D23641466](https://our.internmc.facebook.com/intern/diff/D23641466/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC profiling] Don't wrap toHere() calls with profiling"

server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.

Since `toHere()` does not execute operations (torch operators) over RPC and simply
transfers the value to the local node, we don't need to enable the profiler
remotely for this message. This causes unnecessary overhead and is not needed.

Since `toHere` is a blocking call, we already profile the call on the local node using `RECORD_USER_SCOPE`, so this does not change the expected profiler results (validated by ensuring all remote profiling tests pass).

Differential Revision: [D23641466](https://our.internmc.facebook.com/intern/diff/D23641466/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC profiling] Don't wrap toHere() calls with profiling"

server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.

Since `toHere()` does not execute operations (torch operators) over RPC and simply
transfers the value to the local node, we don't need to enable the profiler
remotely for this message. This causes unnecessary overhead and is not needed.

Since `toHere` is a blocking call, we already profile the call on the local node using `RECORD_USER_SCOPE`, so this does not change the expected profiler results (validated by ensuring all remote profiling tests pass).

Differential Revision: [D23641466](https://our.internmc.facebook.com/intern/diff/D23641466/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC profiling] Don't wrap toHere() calls with profiling"

server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.

Since `toHere()` does not execute operations (torch operators) over RPC and simply
transfers the value to the local node, we don't need to enable the profiler
remotely for this message. This causes unnecessary overhead and is not needed.

Since `toHere` is a blocking call, we already profile the call on the local node using `RECORD_USER_SCOPE`, so this does not change the expected profiler results (validated by ensuring all remote profiling tests pass).

Differential Revision: [D23641466](https://our.internmc.facebook.com/intern/diff/D23641466/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC profiling] Don't wrap toHere() calls with profiling"

server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.

Since `toHere()` does not execute operations (torch operators) over RPC and simply
transfers the value to the local node, we don't need to enable the profiler
remotely for this message. This causes unnecessary overhead and is not needed.

Since `toHere` is a blocking call, we already profile the call on the local node using `RECORD_USER_SCOPE`, so this does not change the expected profiler results (validated by ensuring all remote profiling tests pass).

Differential Revision: [D23641466](https://our.internmc.facebook.com/intern/diff/D23641466/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC profiling] Don't wrap toHere() calls with profiling"

server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.

Since `toHere()` does not execute operations (torch operators) over RPC and simply
transfers the value to the local node, we don't need to enable the profiler
remotely for this message. This causes unnecessary overhead and is not needed.

Since `toHere` is a blocking call, we already profile the call on the local node using `RECORD_USER_SCOPE`, so this does not change the expected profiler results (validated by ensuring all remote profiling tests pass).

Differential Revision: [D23641466](https://our.internmc.facebook.com/intern/diff/D23641466/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC profiling] Don't wrap toHere() calls with profiling"

server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.
server
* #44664 [RPC profiling] Extend RPC profiling to support async function execution over RPC.
* **#44655 [RPC profiling] Don't wrap toHere() calls with profiling**
* #44653 [RPC profiling] Allow disableProfiler() to be called from another thread.
* #44646 Remove thread_local RecordFunctionGuard from profiler.

Since `toHere()` does not execute operations (torch operators) over RPC and simply
transfers the value to the local node, we don't need to enable the profiler
remotely for this message. This causes unnecessary overhead and is not needed.

Since `toHere` is a blocking call, we already profile the call on the local node using `RECORD_USER_SCOPE`, so this does not change the expected profiler results (validated by ensuring all remote profiling tests pass).

Differential Revision: [D23641466](https://our.internmc.facebook.com/intern/diff/D23641466/)

[ghstack-poisoned]
